instrumentation/ratpack-1.4/library/src/test/groovy/io/opentelemetry/instrumentation/ratpack/server/RatpackServerApplicationTest.groovy

```diff
@@ -0,0 +1,118 @@
+import io.opentelemetry.sdk.trace.SdkTracerProvider
+import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor
+import io.opentelemetry.sdk.trace.export.SpanExporter
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes
+import ratpack.exec.ExecInterceptor
+import ratpack.guice.Guice
+import ratpack.handling.Handler
+import ratpack.server.RatpackServer
+import spock.lang.Specification
+import spock.util.concurrent.PollingConditions
+
+import javax.inject.Singleton
+
+class RatpackServerApplicationTest extends Specification {
+
+  def app = new RatpackFunctionalTest(RatpackApp)
+
+  def "add span on handlers"() {
+    expect:
+    app.test { httpClient -> "hi-foo" == httpClient.get("foo").body.text }
+
+    new PollingConditions().eventually {
+      def spanData = app.spanExporter.finishedSpanItems.find { it.name == "/foo" }
+      def attributes = spanData.attributes.asMap()
+
+      spanData.kind == SpanKind.SERVER
+      attributes[SemanticAttributes.HTTP_ROUTE] == "/foo"
+      attributes[SemanticAttributes.HTTP_TARGET] == "/foo"
+      attributes[SemanticAttributes.HTTP_METHOD] == "GET"
+      attributes[SemanticAttributes.HTTP_STATUS_CODE] == 200L
+    }
+  }
+
+  def "ignore handlers before OpenTelemetryServerHandler"() {
+    expect:
+    app.test { httpClient -> "ignored" == httpClient.get("ignore").body.text }
+
+    new PollingConditions(initialDelay: 0.1, timeout: 0.3).eventually {
+      !app.spanExporter.finishedSpanItems.any { it.name == "/ignore" }
+    }
+  }
+}
+
+
+@CompileStatic
+class OpenTelemetryModule extends AbstractModule {
+  @Override
+  protected void configure() {
+    bind(SpanExporter).toInstance(InMemorySpanExporter.create())
+  }
+
+  @Singleton
+  @Provides
+  RatpackTracing ratpackTracing(OpenTelemetry openTelemetry) {
+    return RatpackTracing.create(openTelemetry)
+  }
+
+  @Singleton
+  @Provides
+  Handler ratpackServerHandler(RatpackTracing ratpackTracing) {
+    return ratpackTracing.getOpenTelemetryServerHandler()
+  }
+
+  @Singleton
+  @Provides
+  ExecInterceptor ratpackExecInterceptor(RatpackTracing ratpackTracing) {
+    return ratpackTracing.getOpenTelemetryExecInterceptor()
+  }
+
+  @Provides
+  @Singleton
+  OpenTelemetry providesOpenTelemetry(SpanExporter spanExporter) {
+    def tracerProvider = SdkTracerProvider.builder()
+      .addSpanProcessor(SimpleSpanProcessor.create(spanExporter))
+      .build()
+    return OpenTelemetrySdk.builder().setTracerProvider(tracerProvider).build()
+  }
+}
+
+@CompileStatic
+class RatpackApp {
+  static void main(String... args) {
+    RatpackServer.start { server ->
+      server
+        .registry(
+          Guice.registry { bindings ->
+            bindings
+              .module(OpenTelemetryModule)
+          }
+        )
+        .handlers { chain ->
+          chain
+            .get("ignore") { ctx -> ctx.render("ignored") }
+            .all {
+              it.insert(it.get(RatpackTracing).getOpenTelemetryServerHandler())
```

Comment:
From what I understand, the desire for exposing the `OpenTelemetryServerHandler` class is to be able to reference it directly from the Guice DI context here. Is that correct?

I'm curious whether it's valid to have the handler both in the DI context and here - won't the handler get applied twice?

Reply:
No, it will not.
The handler need to be instanciated in Guice DI context (aka Ratpack registry), but the Handler needs to be added to the Ratpack Chain.

```it.get(RatpackTracing).getOpenTelemetryServerHandler()``` gets `RatpackTracing` from the Registry and access the Handler instance

Reply:
Ok - I think to allow usage through DI like you want it makes sense to make the classes, but not the constructor, public. So we could have the classes and the current getters that return the concrete type. Does that sound ok?

Reply:
That sounds great @anuraaga . I think it makes sense becauase then we enforce to construct them via `RatpackTracing`  🙇

Reply:
@anuraaga applied the changes, now binding the Handler in Ratpack Chain is nicer but the constructions is enforced via `RatpackTracing `